src/parse.yy

```diff
@@ -2813,53 +2819,70 @@ lAddDeclaration(DeclSpecs *ds, Declarator *decl) {
 }
 
 static void
-lAddTemplateDeclaration(TemplateParms *templateParmList, DeclSpecs *ds, Declarator *decl) {
-    if (ds == nullptr || decl == nullptr)
-        // Error happened earlier during parsing
-        return;
-
-    decl->InitFromDeclSpecs(ds);
-    if (ds->storageClass == SC_TYPEDEF) {
-        // FIXME: source position
-        Error(decl->pos, "Illegal \"typedef\" provided with function template.");
-        return;
+lCheckTemplateDeclSpecs(DeclSpecs *ds, SourcePos pos, TemplateType type, const char* name) {
+    std::string templateTypeStr;
+    switch (type) {
+        case TemplateType::Template:
+            templateTypeStr = "function template";
+            break;
+        case TemplateType::Instantiation:
+            templateTypeStr = "template instantiation";
+            break;
+        case TemplateType::Specialization:
+            templateTypeStr = "template specialization";
+            break;
+        default:
+            templateTypeStr = "function template";
+            break;
     }
     if (ds->typeQualifiers & TYPEQUAL_TASK){
-        Error(decl->pos, "'task' not supported for templates.");
+        Error(pos, "'task' not supported for %s.", templateTypeStr.c_str());
         return;
     }
     if (ds->typeQualifiers & TYPEQUAL_EXPORT) {
-        Error(decl->pos, "'export' not supported for templates.");
+        Error(pos, "'export' not supported for %s.", templateTypeStr.c_str());
+        return;
+    }
+    if (ds->storageClass == SC_TYPEDEF) {
+        // FIXME: source position
+        Error(pos, "Illegal \"typedef\" provided with %s.", templateTypeStr.c_str());
         return;
     }
     // We can't support extern "C"/extern "SYCL" for templates because
     // we need mangling information.
     if (ds->storageClass == SC_EXTERN_C || ds->storageClass == SC_EXTERN_SYCL) {
-        Error(decl->pos, "Illegal linkage provided with function template.");
+        Error(pos, "Illegal linkage provided with %s.", templateTypeStr.c_str());
         return;
     }
+    Assert(ds->storageClass == SC_NONE || ds->storageClass == SC_STATIC || ds->storageClass == SC_EXTERN);
+    bool isVectorCall = (ds->typeQualifiers & TYPEQUAL_VECTORCALL);
+    if (isVectorCall) {
+        Error(pos, "Illegal to use \"__vectorcall\" qualifier on non-extern function \"%s\".", name);
+    }
+    bool isRegCall = (ds->typeQualifiers & TYPEQUAL_REGCALL);
+    if (isRegCall) {
+        Error(pos, "Illegal to use \"__regcall\" qualifier on non-extern function \"%s\".", name);
+    }
+}
+
+static void
+lAddTemplateDeclaration(TemplateParms *templateParmList, DeclSpecs *ds, Declarator *decl) {
+    if (ds == nullptr || decl == nullptr)
```

Comment:
Add `{}`. It's already more than one line with the comment, easy to add something and do no notice missing `{}`.

I see that's coming from existing code, but nevertheless.